[Build] Add short compressed build description (#3262)

TD-er opened the pull request:
See #3262

Outputs `Tffff:e000:Pe800:24:4400:(3):800:`
Equivalent to:

```
    "CONTROLLER_SET_ALL",
    "NOTIFIER_SET_NONE",
    "PLUGIN_SET_ONLY_SWITCH",
    "USES_P001",  # Switch
    "USES_P002",  # ADC
    "USES_P004",  # Dallas DS18b20
    "USES_P100",  # Pulse Counter - DS2423
    "USES_C016",  # Cache Controller
    "USES_C018",  # TTN/RN2483
    "USES_C015",  # TTN/RN2483
```

Still missing Notifier, CPU, flash size.

tonhuisman commented:
I think this would also need a Python counterpart, to define the filename before the compile is started?

tonhuisman commented:
Wouldn't it be more natural to first list the 'P' plugins bitmap, then the 'T' controllers bitmap and then the 'N' notifications bitmap? (to be implemented). I *think* most users will first look at the plugins before looking at the controllers, but that may just be me.

TD-er commented:
In size it wouldn't matter and in decoding probably also makes no difference.
But I doubt someone will try to read the filename encoding.
So we should make some decoder where it is most practical.

Maybe a JS piece of code somewhere online where a user can decode the string into readable form.

tonhuisman commented:
Been chewing a bit in the back of my mind for a couple of weeks now, and it feels more and more like a way to go forward. It may need a way to include the 'other' compile-time flags like `FEATURE_SD`, `FEATURE_I2CMULTIPLEXER`, `FEATURE_ADD_VCC` etc. Maybe some of the .h files (`custom.h` , `define_plugin_sets.h`) should be parsed to determine the actual features that will be included, not forgetting the features set from the .py scripts. Found [pypreprocessor](https://github.com/interpreters/pypreprocessor) that could possibly handle that (after some adjustments to make it C-preprocessor compatible, as it now has its own dialect, and doesn't yet support `#if defined(XX)`, comparisons, etc.).

TD-er committed:
[Build] Add short compressed build description (#3262)

See #3262

Outputs `Tffff:7:P17:2400:22:(3):10:`
Equivalent to:

```
    "CONTROLLER_SET_ALL",
    "NOTIFIER_SET_NONE",
    "PLUGIN_SET_ONLY_SWITCH",
    "USES_P001",  # Switch
    "USES_P002",  # ADC
    "USES_P004",  # Dallas DS18b20
    "USES_P100",  # Pulse Counter - DS2423
    "USES_C016",  # Cache Controller
    "USES_C018",  # TTN/RN2483
    "USES_C015",  # TTN/RN2483
```

Still missing Notifier, CPU, flash size.
[Build] Fix bit order in build description


Fix build issues after merge

TD-er committed:
Merge branch 'mega' into feature/build_description

TD-er committed:
Merge branch 'mega' into feature/build_description

TD-er committed:
Merge branch 'mega' into feature/build_description

TD-er committed:
Merge branch 'mega' into feature/build_description

TD-er committed:
Merge branch 'mega' into feature/build_description

TD-er committed:
Merge branch 'mega' into feature/build_description

TD-er committed:
Merge branch 'mega' into feature/build_description

TD-er committed:
Merge branch 'mega' into feature/build_description

TD-er committed:
Merge branch 'mega' into feature/build_description

TD-er committed:
Merge branch 'mega' into feature/build_description

TD-er committed:
Merge branch 'mega' into feature/build_description

TD-er committed:
Merge branch 'mega' into feature/build_description

TD-er committed:
Merge branch 'mega' into feature/build_description

TD-er committed:
Merge branch 'mega' into feature/build_description

TD-er committed:
Merge branch 'mega' into feature/build_description

TD-er committed:
Fix build issues due to changes in ESPEasy core.

TD-er committed:
Merge branch 'mega' into feature/build_description

TD-er committed:
Merge branch 'mega' into feature/build_description

TD-er committed:
Merge branch 'mega' into feature/build_description

TD-er committed:
Merge branch 'mega' into feature/build_description

TD-er committed:
Merge branch 'mega' into feature/build_description

TD-er committed:
Fix missing relative path in include

TD-er committed:
Merge branch 'mega' into feature/build_description

TD-er committed:
Merge branch 'mega' into feature/build_description

TD-er committed:
Merge branch 'mega' into feature/build_description

TD-er committed:
[Build descriptop] Fix merge issue

TD-er committed:
Merge branch 'mega' into feature/build_description

TD-er committed:
Merge branch 'mega' into feature/build_description

TD-er committed:
Merge branch 'mega' into feature/build_description

TD-er commented:
Nee to add a bitmap too for all the `FEATURE_xxx` defines.

tonhuisman commented:
These are the (global) FEATURE_(flags) I found after converting all non-plugin, non-controller, non-Notifier USE_/USES_ settings:
```
Feature flags: (Alphabetic order)
---------------------------------
FEATURE_ADC_VCC
FEATURE_ARDUINO_OTA
FEATURE_BLYNK
FEATURE_CHART_JS
FEATURE_CUSTOM_PROVISIONING
FEATURE_DNS_SERVER
FEATURE_DOMOTICZ
FEATURE_DOWNLOAD
FEATURE_ESPEASY_P2P
FEATURE_ETHERNET
FEATURE_ETHERNET
FEATURE_EXT_RTC
FEATURE_FHEM
FEATURE_HOMEASSISTANT_OPENHAB
FEATURE_I2CMULTIPLEXER
FEATURE_I2C_DEVICE_SCAN
FEATURE_MDNS
FEATURE_MODBUS
FEATURE_MQTT
FEATURE_NON_STANDARD_24_TASKS
FEATURE_NOTIFIER
FEATURE_PACKED_RAW_DATA
FEATURE_PLUGIN_STATS
FEATURE_RTTTL
FEATURE_SD
FEATURE_SERVO
FEATURE_SETTINGS_ARCHIVE
FEATURE_SSDP
FEATURE_TIMING_STATS
FEATURE_TOOLTIPS
FEATURE_TRIGONOMETRIC_FUNCTIONS_RULES
// FEATURE_REPORTING
```

TD-er committed:
Merge branch 'mega' into feature/build_description

TD-er committed:
Merge branch 'mega' into feature/build_description
